framework/sandstone.cpp

```diff
@@ -3250,52 +3394,68 @@ int main(int argc, char **argv)
+                tc = 0;
+                logging_print_iteration_start();
+                initialize_smi_counts();  // used by smi_count test
+            }
 
-        wait_delay_between_tests(restarting);
+            if(sApp->service_background_scan == false)
+                wait_delay_between_tests(restarting);
 
-        struct test *test = get_next_test(tc);
-        if (test == nullptr) break;
+            struct test *test = get_next_test(tc);
+            
+            if (test == nullptr) break;
 
-        // Note temporal coupling here with restarting
-        // it is assigned here but used above in the next iteration
-        restarting = (test == RESTART_OF_TESTS);
+            // Note temporal coupling here with restarting
+            // it is assigned here but used above in the next iteration
+            restarting = (test == RESTART_OF_TESTS);
 
-        if (restarting) {
-            if constexpr (InterruptMonitor::InterruptMonitorWorks) {
-                if (mce_test.quality_level != TEST_QUALITY_SKIP)
-                    test = &mce_test;
+            if (restarting) {
+                if constexpr (InterruptMonitor::InterruptMonitorWorks) {
+                    if (mce_test.quality_level != TEST_QUALITY_SKIP)
+                        test = &mce_test;
+                }
+                if (test == RESTART_OF_TESTS)
+                    continue;
             }
-            if (test == RESTART_OF_TESTS)
-                continue;
-        }
 
-        TestResult r = run_one_test(&tc, test, per_cpu_failures);
-        if (r == TestFailed)
-            ++total_failures;
-        else if (r == TestPassed)
-            ++total_successes;
-        else if (r == TestSkipped)
-            ++total_skips;
-        total_tests_run++;
+            TestResult r = run_one_test(&tc, test, per_cpu_failures);
 
-        // keep the record of failures to triage later
-        if (total_failures > triage_tests.size()) {
-            // ### use per_cpu_fails??
-            triage_tests.push_back(test);
-        }
+            if(sApp->service_background_scan == true)
+                background_scan_update_timestamps();
 
-        if (total_tests_run >= sApp->max_test_count)
-            break;
-    }
+            if (r == TestFailed)
+                ++total_failures;
+            else if (r == TestPassed)
+                ++total_successes;
+            else if (r == TestSkipped)
+                ++total_skips;
+            
+            total_tests_run++;
+
+            // keep the record of failures to triage later
+            if (total_failures > triage_tests.size()) {
+                // ### use per_cpu_fails??
+                triage_tests.push_back(test);
+            }
+
+            if (total_tests_run >= sApp->max_test_count)
+                break;
+
+            if(sApp->service_background_scan == true)
+                background_scan_daemon();
+        }
+    } while(sApp->service_background_scan == true);
```

Comment:
The variable is never reset back to `false`, so we'll run forever even in the case of failures detected. I don't think we want that. Or am I missing something?

Reply:
I checked what's going to happen if a test returns TestFailed and we exit the loop. 
However, you raised an important issue here which reminds me to check what's going to happen when we go over the whole test list.

Reply:
But what happens when the inner loop exits due to a failed test? Do we mean to continue running? I think that is in conflict with `fatal_errors` and even `fatal_skips`.

I don't even understand what you're trying to accomplish here. If you meant to run forever, then I think the user should pass `-T forever` in the command-line (and we can enforce that in the restricted command-line if we want to).

Since the condition in the inner `while` is going to become complex, I recommend:
```c++
    bool should_run_more_tests;
    do {
        // other code goes here

        ++total_tests_run;
        should_run_more_tests = (total_tests_run < sApp->max_test_count);
        if (r == TestFailed) {
            ++total_failures;
            should_run_more_tests = should_run_more_tests && !fatal_errors;
        } else if (r == TestPassed) {
            ++total_successes;
        } else if (r == TestSkipped) {
            ++total_skips;
            should_run_more_tests = should_run_more_tests && !fatal_skips;
        }
    } while (should_run_more_tests);
```

Reply:
Let me explain what I wanted to achieve:

I thought we want the background scan to go over the whole test list and if no errors are found just start a next pass when the time is right. Running a background scan which ends at the end of test list doesn't make sense to me.
If we want that we should return something to tell the service has ended.

I thought we need an external loop to have this. However, now I think I was incorrect here.

I dug more into the code and I see we already have this:

```
        case service_option:
            fatal_errors = true;
            sApp->endtime = MonotonicTimePoint::max(); <-- here
            sApp->service_background_scan = true;
            break;
```

This should not interfere with test failures or test skips. From what I see the max value should be big enough so id doesn't overflow in reasonable time period.